statistics/update_test.go

```diff
@@ -631,6 +631,87 @@ func (s *testStatsUpdateSuite) TestQueryFeedback(c *C) {
 	c.Assert(h.HandleUpdateStats(s.do.InfoSchema()), IsNil)
 }
 
+func (s *testStatsUpdateSuite) TestQueryFeedbackForPartition(c *C) {
+	defer cleanEnv(c, s.store, s.do)
+	testKit := testkit.NewTestKit(c, s.store)
+	testKit.MustExec("use test")
+	testKit.MustExec("set @@session.tidb_enable_table_partition=1")
+	testKit.MustExec(`create table t (a bigint(64), b bigint(64), primary key(a), index idx(b))
+			    partition by range (a) (
+			    partition p0 values less than (3),
+			    partition p1 values less than (6))`)
+	testKit.MustExec("insert into t values (1,2),(2,2),(3,4),(4,1),(5,6)")
+	testKit.MustExec("analyze table t")
+
+	oriProbability := statistics.FeedbackProbability
+	oriNumber := statistics.MaxNumberOfRanges
+	defer func() {
+		statistics.FeedbackProbability = oriProbability
+		statistics.MaxNumberOfRanges = oriNumber
+	}()
+	h := s.do.StatsHandle()
+	statistics.FeedbackProbability = 1
+	tests := []struct {
+		sql     string
+		hist    string
+		idxCols int
+	}{
+		{
+			// test primary key feedback
+			sql: "select * from t where t.a <= 5",
+			hist: "column:1 ndv:2 totColSize:0\n" +
+				"num: 1\tlower_bound: 1\tupper_bound: 1\trepeats: 1\n" +
+				"num: 2\tlower_bound: 2\tupper_bound: 2\trepeats: 1",
+			idxCols: 0,
+		},
+		{
+			// test index feedback by double read
+			sql: "select * from t use index(idx) where t.b <= 5",
+			hist: "index:1 ndv:1\n" +
+				"num: 2\tlower_bound: 2\tupper_bound: 2\trepeats: 2",
+			idxCols: 1,
+		},
+		{
+			// test index feedback by single read
+			sql: "select b from t use index(idx) where t.b <= 5",
+			hist: "index:1 ndv:1\n" +
+				"num: 2\tlower_bound: 2\tupper_bound: 2\trepeats: 2",
+			idxCols: 1,
+		},
+	}
+	is := s.do.InfoSchema()
+	table, err := is.TableByName(model.NewCIStr("test"), model.NewCIStr("t"))
+	c.Assert(err, IsNil)
+	tblInfo := table.Meta()
+	pi := tblInfo.GetPartitionInfo()
+	c.Assert(pi, NotNil)
+
+	// This test will check the result of partition p0.
+	var pid int64
+	for _, def := range pi.Definitions {
+		if def.Name.L == "p0" {
+			pid = def.ID
+			break
+		}
+	}
+
+	for i, t := range tests {
+		testKit.MustQuery(t.sql)
+		c.Assert(h.DumpStatsDeltaToKV(statistics.DumpAll), IsNil)
+		c.Assert(h.DumpStatsFeedbackToKV(), IsNil)
+		c.Assert(h.HandleUpdateStats(s.do.InfoSchema()), IsNil)
```

Comment:
The stats is not actually updated, and it should not be updated, because there is no change to the table since last analyze. For feedback to fully work, there is still something to be done in the stats package.

Reply:
Done.